### PR TITLE
feat: expose device token

### PR DESCRIPTION
**Pull request by Shahroz16**
Complete each step to get your pull request merged in. [Learn more about the workflow this project uses](https://github.com/customerio/customerio-android/develop/docs/dev-notes/GIT-WORKFLOW.md).
- [ ] [Assign members of your team](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to review the pull request.
- [ ] Wait for pull request status checks to complete. If there are problems, fix them until you see that [all status checks are passing](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fsymfony.com%2Fdoc%2F4.3%2F_images%2Fdocs-pull-request-symfonycloud.png&f=1&nofb=1).
- [ ] Wait until the pull request has been reviewed *and approved* by a teammate
- [ ] After pull request is approved, and you determine it's ready **add the label "Ready to merge"** to the pull request and it will automatically be merged.

### CODE REVIEW

**Comment by mrehan27** [comment]
@Shahroz16 I'm not concerned about the sample app, it will work fine for it. I'm here asking for cases where customers need to utilize token for other SDK (e.g. [tracking installs](https://github.com/customerio/customerio-reactnative/issues/127#issuecomment-1535041929), etc.), generally they will be fetching the token at app start. Assuming network and everything else is working fine, do you think the token will be available every time at first launch as well? If yes, then I'm okay with this, if not, I would recommend to make it async.

**Comment by Shahroz16** [comment]
@mrehan27 

>generally they will be fetching the token at app start 

I don't think that's always true, because I believe it is usually done after onboarding/sign-up. The adjust library also doesn't need it for the app starts but anytime in the app lifecycle. 

Additionally, we don't generate token in iOS we are reliant on Apps providing us that information, so in that case, if we make this call async for android, and sync for iOS there will be a disparity.

**Comment by mrehan27** [comment]
@Shahroz16 I think it strongly depends on the product, we cannot assume if it has to be done at the start of onboarding or after completion.

For the async part, we can always make it async ourselves and returns the results wrapped in future/promise instantly. The original ticket was also [updated as async design](https://github.com/customerio/issues/issues/10298#issuecomment-1578801133) to avoid any changes to this later.

I still think having it async is more flexible for cases when we are fetching directly from FCM. But if you think the current solution will work fine for customers, I'm happy releasing changes without making it async.

File: sdk/src/sharedTest/java/io/customer/sdk/CustomerIOTest.kt [hunk 1]
```diff
@@ -103,6 +103,17 @@ class CustomerIOTest : BaseTest() {
         updatedConfig.trackingApiHostname shouldBeEqualTo givenTrackingApiUrl
     }
 
+    @Test
+    fun deviceToken_givenGetValue_expectDeviceRepositoryGetDeviceToken() {
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/CustomerIOTest.kt:107]: Could we add a test when there hasn't been been a device token set yet? Where you expect to get `null`? 

Also, a non-blocking idea to share, I think that this feature benefit better from integration tests. 

```kotlin
assertNull(CustomerIO.instance().deviceToken)

CustomerIO.registerDeviceToken(givenToken)

assertEqual(CustomerIO.instance().deviceToken, givenToken)
```

This feels closer to the way customers will use it so I think it's a higher quality test.

File: sdk/src/main/java/io/customer/sdk/CustomerIO.kt [hunk 2]
```diff
@@ -538,6 +540,9 @@ class CustomerIO internal constructor(
             deviceRepository.addCustomDeviceAttributes(value)
         }
 
+    override val deviceToken: String?
```
Comment (levibostian) [inline sdk/src/main/java/io/customer/sdk/CustomerIO.kt:543]: I thought we were going to use `currentlyRegisteredDeviceToken`? Or some other name that helps communicate that the value we return to you is one that you have registered with the SDK. `registeredDeviceToken` could be good, too.

**Review by levibostian — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -28,6 +28,10 @@ lint:
 lint-install:
 	./scripts/get-ktlint.sh
 
+# Generate public API binary
+generate-api:
+	./gradlew apiDump
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply Makefile:33]: This generates the files, and the binary validator runs, and validates. So, previously we exposed `deviceToken` and now we changed it, it would have been a breaking change. So when I merged it, it gave me an error saying i am changing a publically exposed `API`. 

```
public fun getRegisteredDeviceToken ()Ljava/lang/String;
```

the change this command has created and is part of out `.api` file. Next time, someone changes it, it will match with this file and give an error.

[the remark this answers is not in the document]
Comment (levibostian) [reply Makefile:33]: Oh, ok. 

You did good at explaining this. The names of the commands and the comments inside of the Makefile don't do as good of a job, IMO. 

A suggestion is changing the name of the commands in the Makefile:

```
generate-public-api: 
   ...

validate-public-api:
   ...
```

The way the file is now, I don't know what command to use, when. If commands explained the purpose, it could explain that you should run `generate-public-api` while developing a feature and run `validate-public-api` on CI to verify the stability of public-facing API.
